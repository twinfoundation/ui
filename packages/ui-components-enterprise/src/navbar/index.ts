// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

export { Navbar } from "./navbar";
export type * from "./navbarProps";
