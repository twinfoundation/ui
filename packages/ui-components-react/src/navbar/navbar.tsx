// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
import {
	Navbar as FlowbiteNavbar,
	NavbarBrand,
	NavbarCollapse,
	NavbarLink,
	NavbarToggle
} from "flowbite-react";
import { memo, type JSX } from "react";
import type { NavbarProps } from "./navbarProps";

/**
 * A responsive navigation bar component that adapts to different screen sizes.
 * Provides a flexible container for branding, navigation links, and other content.
 * @example
 * ```tsx
 * <Navbar fluid rounded>
 *   <Navbar.Brand href="https://example.com">
 *     <img src="/logo.svg" className="mr-3 h-6 sm:h-9" alt="Logo" />
 *     <span className="self-center whitespace-nowrap text-xl font-semibold">Brand</span>
 *   </Navbar.Brand>
 *   <Navbar.Toggle />
 *   <Navbar.Collapse>
 *     <Navbar.Link href="#" active>Home</Navbar.Link>
 *     <Navbar.Link href="#">About</Navbar.Link>
 *     <Navbar.Link href="#">Services</Navbar.Link>
 *   </Navbar.Collapse>
 * </Navbar>
 * ```
 */
const NavbarComponent = memo(
	({ children, fluid = false, rounded = false, ...props }: NavbarProps): JSX.Element => (
		<FlowbiteNavbar fluid={fluid} rounded={rounded} {...props}>
			{children}
		</FlowbiteNavbar>
	)
);

NavbarComponent.displayName = "Navbar";

export const Navbar = Object.assign(NavbarComponent, {
	Brand: NavbarBrand,
	Collapse: NavbarCollapse,
	Link: NavbarLink,
	Toggle: NavbarToggle
});
