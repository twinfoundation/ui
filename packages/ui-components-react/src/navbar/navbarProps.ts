// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.
import type { InferProps } from "prop-types";
import type { PropsWithChildren } from "react";

export const NavbarPropTypes = {};

/**
 * Navbar props.
 */
export type NavbarProps = PropsWithChildren<InferProps<typeof NavbarPropTypes>>;
