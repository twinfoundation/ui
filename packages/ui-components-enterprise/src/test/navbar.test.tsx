// Copyright 2024 IOTA Stiftung.
// SPDX-License-Identifier: Apache-2.0.

import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { Navbar } from '../navbar/navbar';

describe('Navbar', () => {
  it('renders basic navbar correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <span>Brand</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with fluid prop correctly', () => {
    const { container } = render(
      <Navbar fluid>
        <Navbar.Brand href="/">
          <span>Fluid Brand</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with rounded prop correctly', () => {
    const { container } = render(
      <Navbar rounded>
        <Navbar.Brand href="/">
          <span>Rounded Brand</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with both fluid and rounded props correctly', () => {
    const { container } = render(
      <Navbar fluid rounded>
        <Navbar.Brand href="/">
          <span>Fluid Rounded Brand</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with brand and links correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <span>Brand</span>
        </Navbar.Brand>
        <Navbar.Collapse>
          <Navbar.Link href="/" active>
            Home
          </Navbar.Link>
          <Navbar.Link href="/about">About</Navbar.Link>
          <Navbar.Link href="/contact">Contact</Navbar.Link>
        </Navbar.Collapse>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with toggle button correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <span>Brand</span>
        </Navbar.Brand>
        <Navbar.Toggle />
        <Navbar.Collapse>
          <Navbar.Link href="/">Home</Navbar.Link>
        </Navbar.Collapse>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with custom className correctly', () => {
    const { container } = render(
      <Navbar className="custom-navbar">
        <Navbar.Brand href="/">
          <span>Custom Brand</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with active link correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <span>Brand</span>
        </Navbar.Brand>
        <Navbar.Collapse>
          <Navbar.Link href="/" active>
            Active Link
          </Navbar.Link>
          <Navbar.Link href="/inactive">Inactive Link</Navbar.Link>
        </Navbar.Collapse>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with complex brand content correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <img src="/logo.svg" alt="Logo" className="h-6 w-6" />
          <span className="ml-2 text-xl font-bold">Company Name</span>
        </Navbar.Brand>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders navbar with multiple links correctly', () => {
    const { container } = render(
      <Navbar>
        <Navbar.Brand href="/">
          <span>Brand</span>
        </Navbar.Brand>
        <Navbar.Collapse>
          <Navbar.Link href="/">Home</Navbar.Link>
          <Navbar.Link href="/products">Products</Navbar.Link>
          <Navbar.Link href="/services">Services</Navbar.Link>
          <Navbar.Link href="/about">About</Navbar.Link>
          <Navbar.Link href="/contact">Contact</Navbar.Link>
        </Navbar.Collapse>
      </Navbar>
    );
    expect(container).toMatchSnapshot();
  });
}); 