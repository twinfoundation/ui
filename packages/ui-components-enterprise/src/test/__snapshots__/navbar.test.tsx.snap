// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > renders basic navbar correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Brand
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with active link correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Brand
        </span>
      </a>
      <div
        class="w-full md:block md:w-auto hidden"
        data-testid="flowbite-navbar-collapse"
      >
        <ul
          class="mt-4 flex flex-col md:mt-0 md:flex-row md:space-x-8 md:text-sm md:font-medium"
        >
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 bg-cyan-700 text-white dark:text-white md:bg-transparent md:text-cyan-700"
              href="/"
            >
              Active Link
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/inactive"
            >
              Inactive Link
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with both fluid and rounded props correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4 rounded"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Fluid Rounded Brand
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with brand and links correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Brand
        </span>
      </a>
      <div
        class="w-full md:block md:w-auto hidden"
        data-testid="flowbite-navbar-collapse"
      >
        <ul
          class="mt-4 flex flex-col md:mt-0 md:flex-row md:space-x-8 md:text-sm md:font-medium"
        >
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 bg-cyan-700 text-white dark:text-white md:bg-transparent md:text-cyan-700"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/about"
            >
              About
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/contact"
            >
              Contact
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with complex brand content correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <img
          alt="Logo"
          class="h-6 w-6"
          src="/logo.svg"
        />
        <span
          class="ml-2 text-xl font-bold"
        >
          Company Name
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with custom className correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4 custom-navbar"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Custom Brand
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with fluid prop correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Fluid Brand
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with multiple links correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Brand
        </span>
      </a>
      <div
        class="w-full md:block md:w-auto hidden"
        data-testid="flowbite-navbar-collapse"
      >
        <ul
          class="mt-4 flex flex-col md:mt-0 md:flex-row md:space-x-8 md:text-sm md:font-medium"
        >
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/products"
            >
              Products
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/services"
            >
              Services
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/about"
            >
              About
            </a>
          </li>
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/contact"
            >
              Contact
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with rounded prop correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4 rounded"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Rounded Brand
        </span>
      </a>
    </div>
  </nav>
</div>
`;

exports[`Navbar > renders navbar with toggle button correctly 1`] = `
<div>
  <nav
    class="bg-white px-2 py-2.5 dark:border-gray-700 dark:bg-gray-800 sm:px-4"
  >
    <div
      class="mx-auto flex flex-wrap items-center justify-between container"
    >
      <a
        class="flex items-center"
        href="/"
      >
        <span>
          Brand
        </span>
      </a>
      <button
        class="inline-flex items-center rounded-lg p-2 text-sm text-gray-500 hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 dark:hover:bg-gray-700 dark:focus:ring-gray-600 md:hidden"
        data-testid="flowbite-navbar-toggle"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6 shrink-0"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          />
        </svg>
      </button>
      <div
        class="w-full md:block md:w-auto hidden"
        data-testid="flowbite-navbar-collapse"
      >
        <ul
          class="mt-4 flex flex-col md:mt-0 md:flex-row md:space-x-8 md:text-sm md:font-medium"
        >
          <li>
            <a
              class="block py-2 pl-3 pr-4 md:p-0 border-b border-gray-100 text-gray-700 hover:bg-gray-50 dark:border-gray-700 dark:text-gray-400 dark:hover:bg-gray-700 dark:hover:text-white md:border-0 md:hover:bg-transparent md:hover:text-cyan-700 md:dark:hover:bg-transparent md:dark:hover:text-white"
              href="/"
            >
              Home
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
